api/internal/git/repospec.go

```diff
@@ -227,56 +227,111 @@ func parsePath(n string) string {
+		return normalizedGithubHost(scheme, username), rest
+	}
+
+	sepIndex := findPathSeparator(n, username == gitUsername)
+	host, rest := n[:sepIndex+1], n[sepIndex+1:]
+
+	if !validHostSpecParsed(scheme, username, host) {
+		return "", n
+	}
+	return scheme + username + host, rest
+}
+
+func validHostSpecParsed(scheme, username, host string) bool {
+	isSCP := username == gitUsername
+	return len(host) > 0 &&
+		(scheme != "" || isSCP) // scheme may be blank for URLs like git@github.com:org/repo
+}
+
+func extractScheme(s string) (string, string) {
+	for _, prefix := range []string{"gh:", "ssh://", "https://", "http://", "file://"} {
+		if rest, found := trimPrefixIgnoreCase(s, prefix); found {
+			return prefix, rest
 		}
 	}
-	if host == "git@" {
-		i := strings.Index(n, "/")
-		if i > -1 {
-			host += n[:i+1]
-			n = n[i+1:]
-		} else {
-			i = strings.Index(n, ":")
-			if i > -1 {
-				host += n[:i+1]
-				n = n[i+1:]
-			}
+	return "", s
+}
+
+func extractUsername(s string) (string, string) {
+	if trimmed, found := trimPrefixIgnoreCase(s, gitUsername); found {
+		return gitUsername, trimmed
+	}
+	return "", s
+}
+
+func trimGithubHost(s string) (string, bool) {
+	for _, prefix := range []string{"github.com/", "github.com:"} {
+		if trimmed, found := trimPrefixIgnoreCase(s, prefix); found {
+			return trimmed, true
 		}
-		return host, n
 	}
+	return s, false
+}
 
-	// If host is a http(s) or ssh URL, grab the domain part.
-	for _, p := range []string{
-		"ssh://", "https://", "http://"} {
-		if strings.HasSuffix(host, p) {
-			i := strings.Index(n, "/")
-			if i > -1 {
-				host += n[0 : i+1]
-				n = n[i+1:]
-			}
-			break
+// trimPrefixIgnoreCase returns the rest of s and true if prefix, ignoring case, prefixes s.
+// Otherwise, trimPrefixIgnoreCase returns s and false.
+func trimPrefixIgnoreCase(s, prefix string) (string, bool) {
+	if len(prefix) <= len(s) && strings.ToLower(s[:len(prefix)]) == prefix {
+		return s[len(prefix):], true
+	}
+	return s, false
+}
+
+func findPathSeparator(hostPath string, isSCP bool) int {
+	if strings.Contains(hostPath, "://") {
+		// This function assumes that the hostPath does not contain a scheme.
+		// If it does, we return -1 to indicate that the hostPath is invalid.
+		return -1
+	}
+
+	sepIndex := strings.Index(hostPath, "/")
+	if isSCP {
+		// The colon acts as a delimiter for scp protocol only if not prefixed by '/'.
+		if colonIndex := strings.Index(hostPath, ":"); colonIndex > 0 && colonIndex < sepIndex {
+			sepIndex = colonIndex
 		}
 	}
+	return sepIndex
+}
+
+const normalizedHTTPGithub = "https://github.com/"
+const gitUsername = "git@"
+const normalizedSCPGithub = gitUsername + "github.com:"
 
-	return normalizeGitHostSpec(host), n
+func normalizedGithubHost(scheme, username string) string {
+	if strings.HasPrefix(scheme, "ssh://") || username == gitUsername {
+		return normalizedSCPGithub
+	}
+	return normalizedHTTPGithub
 }
 
 func normalizeGitHostSpec(host string) string {
```

Comment:
Concern for future PR: I find `normalizeGitHostSpec` hard to reason about because it checks for target strings out of context (with `strings.Contains`). For example, the  repo url `ssh://git@ssh.github.com:443/YOUR-USERNAME/YOUR-REPOSITORY.git` with the following use case: https://docs.github.com/en/authentication/troubleshooting-ssh/using-ssh-over-the-https-port

would be normalized to `git@github.com/YOUR-USERNAME/YOUR-REPOSITORY.git`, which might come as a surprise to the user.

At the very least, I'm glad we no longer use this function in `extractHost`.

Reply:
Yeah, my next PR is getting rid of this entirely in favour of handling all host parsing with `extractHost`! That's a great example of changed behaviour in this PR though, and I added it as a test.